prov/verbs/src/fi_verbs.c

```diff
@@ -815,22 +793,69 @@ static int fi_ibv_init_info(const struct fi_info *hints)
 	default:
 		FI_INFO(&fi_ibv_prov, FI_LOG_CORE, "Unknown transport type");
 		ret = -FI_ENODATA;
-		goto err3;
+		goto err;
 	}
 
 	if (!(fi->domain_attr->name = strdup(ctx->device->name))) {
 		ret = -FI_ENOMEM;
-		goto err3;
+		goto err;
+	}
+
+	*info = fi;
+	return 0;
+err:
+	fi_freeinfo(fi);
+	return ret;
+}
+
+static int fi_ibv_init_info(const struct fi_info *hints)
+{
+	struct ibv_context **ctx_list;
+	struct fi_info *fi, *fi_head, *fi_prev;
+	int ret, i, num_devices;
+
+	if (verbs_info)
+		return 0;
+
+	pthread_mutex_lock(&verbs_info_lock);
+	if (verbs_info)
+		goto unlock;
+
+	if (!fi_ibv_have_device()) {
+		ret = -FI_ENODATA;
+		goto err1;
 	}
 
-	verbs_info = fi;
+	ctx_list = rdma_get_devices(&num_devices);
+	if (!num_devices) {
+		ret = (errno == ENODEV) ? -FI_ENODATA : -errno;
+		goto err1;
+	}
+
+	fi_head = fi_prev = fi = NULL;
```

Comment:
There's no need to set fi = NULL, we should check the return code to fi_ibv_get_info_ctx below